bittide-extra/src/Numeric/Extra.hs

```diff
@@ -0,0 +1,28 @@
+-- SPDX-FileCopyrightText: 2024 Google LLC
+--
+-- SPDX-License-Identifier: Apache-2.0
+
+module Numeric.Extra where
+
+import Clash.Prelude
+
+import Control.Monad (foldM)
+import Data.Char (digitToInt, isHexDigit)
+
+-- | Parse a hexadecimal string into a 'BitPack'able type.
+--
+-- Note that this function does not handle types that do not use the full range
+-- of their bit size. For example, 'Index' will return an 'XException' if the
+-- parsed value is out of range of the @'Index' n@, but in range of
+-- @'BitVector' ('BitSize' ('Index' n))@. To fix this properly, we need a version
+-- of 'unpack' that returns a 'Maybe' value.
+parseHex :: forall a. BitPack a => String -> Either String a
```

Comment:
This introduces a lot of unnecessary `Integer` conversions. Hence, in order to avoid the confusing `log2` function of the previous version and also handle the empty string, how about

```haskell
import Clash.Prelude (BitPack(..), BitSize, (.|.), natToNum, shift)

import Control.Monad (foldM)
import Data.Char (digitToInt, isHexDigit)

-- | Parse a hexadecimal string into a 'BitPack'able type.
--
-- Note that this function does not handle types that do not use the full range
-- of their bit size. For example, 'Index' will return an 'XException' if the
-- parsed value is out of range of the @'Index' n@, but in range of
-- @'BitVector' ('BitSize' ('Index' n))@. To fix this properly, we need a version
-- of 'unpack' that returns a 'Maybe' value.
parseHex :: forall a. BitPack a => String -> Either String a
parseHex "" = Left "Empty String"
parseHex (dropWhile (== '0') -> s) =
  if   not (null s) && isHexDigit (head s)
    && 4 * (length s - 1) + digitBits (head s) > natToNum @(BitSize a)
  then Left $ "Value is out of range: " <> s
  else unpack <$> foldM parseDigit 0 s
 where
  parseDigit !a c
    | not (isHexDigit c) = Left $ "Non-hexadecimal digit: " <> [c] <> " in " <> s
    | otherwise          = Right $ shift a 4 .|. digitToNum c

  digitBits x = case digitToInt x of
    { 0 -> 0; 1 -> 1; 2 -> 2; 3 -> 2; 4 -> 3; 5 -> 3; 6 -> 3; 7 -> 3; _ -> 4 }

  -- 'digitToInt' produces only 16 different values. Hence, it can be
  --  extended to work for any 'Num' instance.
  digitToNum x = case digitToInt x of
    { 0 -> 0; 1 -> 1; 2 -> 2; 3 -> 3; 4 -> 4; 5 -> 5; 6 -> 6; 7 -> 7; 8 -> 8
    ; 9 -> 9; 10 -> 10; 11 -> 11; 12 -> 12; 13 -> 13; 14 -> 14; 15 -> 15
    ; y | y < 0     -> error "digitToInt returned some negative value"
        | otherwise -> error "digitToInt returned some value greater than 15"
    }

```

Reply:
What integer conversions do you see? I only see this one inside the loop:

```haskell
Right $ shift a 4 .|. toInteger (digitToInt c)
```

(I had a superfluous `fromInteger` in there, woops.) And one outside:

```haskell
fromInteger result
```

Both are _very_ close to being "free". `Int` -> `Integer` is a matter of stripping a constructor [`I#`](https://hackage.haskell.org/package/base-4.19.1.0/docs/GHC-Int.html) and replacing it with [`IS`](https://hackage.haskell.org/package/ghc-bignum-1.3/docs/GHC-Num-Integer.html#v:IS). My guess is that GHC optimizes this to a no-op, because that should be representationally equal. Likewise for `Integer` -> `Natural`, if the integer >= 0 (which is checked), it's just a repack.

Handling the empty string is definitely a good idea, that's something I missed!

Reply:
If you're worried about `Integer` vs `Natural`: yeah, I believe there should be some overhead. Keep in mind we do use two `Naturals` to implement `BitVector` though, so whether it'll be more inefficient than that... I'm not sure.

If we want to make this as "neat" as possible I suggest we include our own `digitToNatural` (inspired by `digitToInt`):

```haskell
digitToNatural :: Char -> Either String Natural
digitToNatural c
  | dec  <= 9 = Right (dec)
  | hexl <= 5 = Right (hexl + 10)
  | hexu <= 5 = Right (hexu + 10)
  | otherwise = Left ("digitToNatural: not a digit " ++ show c)
  where
    dec  = fromIntegral (ord c - ord '0')
    hexl = fromIntegral (ord c - ord 'a')
    hexu = fromIntegral (ord c - ord 'A')
```

which would remove some conversions, make the accumulator type a `Natural` (instead of `Integer`), and eliminate the `isHexDigit` check:

```haskell
  parseDigit !a c =
    d <- digitToNatural c
    pure $ shift a 4 .|. d
```

Reply:
But then you can also use `digitToNum`, as `Natural` has a `Num` instance as well. I actually added that in the first place, because I don't understand why the base libraries restrict this function to `Int` as the extension to any `Num` instance turns out to be quite "natural" as well :stuck_out_tongue_winking_eye:.

And we should avoid copying code, even if it is inspired by the base libraries.

Reply:
Yeah `digitToNum` is better. I also don't know why `Int` is so pervasive in `base`. 

> And we should avoid copying code, even if it is inspired by the base libraries.

I of course agree general, but this also has its limits. We're dealing with a function that has some downsides (its partial and not polymorphic enough) and is very unlikely to change in implementation. So :shrug:.

Reply:
Okay I think we've rabbit holed this discussion enough :-). I'm going to press the merge button :eyes:.

Reply:
(Ignore the accidentally broken commit :woozy_face:, fixed now)